aws-s3-transfer-manager/src/types.rs

```diff
@@ -179,7 +179,20 @@ fn is_hidden(path: &Path) -> bool {
 impl Default for UploadFilter {
     fn default() -> Self {
         Self {
-            predicate: Arc::new(|item| item.metadata().is_file() && !is_hidden(item.path())),
+            predicate: Arc::new(|item| {
+                if is_hidden(item.path()) {
```

Comment:
Good at spotting a hole in the logic :)

> If a user purposefully sets source = "/path/to/.hidden/subfolder_to_upload"

Wrote a dummy unit test and confirmed `subfolder_to_upload` will be yielded as a file to upload to S3. Also confirmed that Java Transfer Manager only supports [`DownloadFilter`](https://github.com/aws/aws-sdk-java-v2/blob/master/services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/config/DownloadFilter.java#L30) but not `UploadFilter` currently.

At least I'll restore your original [TODO](https://github.com/awslabs/aws-s3-transfer-manager-rs/blob/c88e07f7c92a3fe8bd391c624876538743c5314b/aws-s3-transfer-manager/src/operation/upload_objects/builders.rs#L110-L115), move it to `UploadFilter::default()`, and update it to say it does not filter a hidden directory.

> Would it make more sense if we called the UploadFilter on directories to decide whether to enter them?

Yes it would, and it'd be ideal if `WalkDir` could support that out of the box; even if we successfully apply the `UploadFilter` to skip a hidden directory, subsequent calls to `WalkDir::next()` will still yield files from within that hidden directory, so if we could stop the iteration at the source, that'd be great.

If we need to filter out hidden directories in `UploadFilter`, we should avoid repeatedly inspecting every path component of a `DirEntry` to check if it's inside a hidden directory; doing so in each `UploadFilter ` call could negatively impact the performance of the common case. Maintaining a hash set of hidden directories can be tricky as well, especially if we want to parallelize directory traversal for future performance improvements, as we might have to give up the guarantee of directory traversal order.

Reply:
Good at spotting a hole in the logic :)

> If a user purposefully sets source = "/path/to/.hidden/subfolder_to_upload"

Wrote a dummy unit test and confirmed `subfolder_to_upload` will be yielded as a file to upload to S3. Also confirmed that Java Transfer Manager only supports [`DownloadFilter`](https://github.com/aws/aws-sdk-java-v2/blob/master/services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/config/DownloadFilter.java#L30) but not `UploadFilter` currently.

At least I'll add a comment to `UploadFilter::default()` that says it does not filter hidden directories

> Would it make more sense if we called the UploadFilter on directories to decide whether to enter them?

Yes it would, and it'd be ideal if `WalkDir` could support that out of the box; even if we successfully apply the `UploadFilter` to skip a hidden directory, subsequent calls to `WalkDir::next()` will still yield files from within that hidden directory, so if we could stop the iteration at the source, that'd be great.

If we need to filter out hidden directories in `UploadFilter`, we should avoid repeatedly inspecting every path component of a `DirEntry` to check if it's inside a hidden directory; doing so in each `UploadFilter ` call could negatively impact the performance of the common case. Maintaining a hash set of hidden directories can be tricky as well, especially if we want to parallelize directory traversal for future performance improvements, as we might have to give up the guarantee of directory traversal order.